File: src/main/java/org/example/keycloak/schemas/ScimUserResponse.java
package org.example.keycloak.schemas;

import java.time.OffsetDateTime;
import java.util.List;
import org.keycloak.models.UserModel;

public record ScimUserResponse(
    List<String> schemas,

    ScimUserName name,
    List<ScimUserEmail> emails,
    String userName,
    String id,
    boolean active
) {
  public ScimUserResponse(
      UserModel user
  ) {
    this(
        List.of("urn:ietf:params:scim:schemas:core:2.0:User"),
        new ScimUserName(user.getLastName(), user.getFirstName()),
        List.of(new ScimUserEmail(user.getEmail(), true)),
        user.getUsername(),
        user.getId(),
        user.isEnabled()
    );
  }

  public record ScimUserName(
      String familyName,
      String givenName
  ) {

  }

  public record ScimUserEmail(
      String value,
      boolean primary
  ) {

  }

  public record ScimUserMeta(
      String resourceType,
      OffsetDateTime created,
      OffsetDateTime lastModified,
      String location
  ) {

  }
}

File: src/main/kotlin/org/example/keycloak/schemas/ScimUserResponse.kt
package org.example.keycloak.schemas

import org.keycloak.models.KeycloakUriInfo
import org.keycloak.models.RealmModel
import org.keycloak.models.UserModel

data class ScimUserResponse(
    val schemas: List<String>,
    val name: ScimUserName,
    val emails: List<ScimUserEmail>,
    val userName: String,
    val id: String,
    val active: Boolean,
    val meta: ScimUserMeta
) {
    constructor(
        user: UserModel,
        realm: RealmModel,
        uriInfo: KeycloakUriInfo
    ) : this(
        schemas = listOf("urn:ietf:params:scim:schemas:core:2.0:User"),
        name = ScimUserName(user.lastName, user.firstName),
        emails = listOf(ScimUserEmail(user.email, true)),
        userName = user.username,
        id = user.id,
        active = user.isEnabled,
        meta = ScimUserMeta(
            resourceType = "User",
            location = uriInfo.baseUriBuilder
                .path("realms")
                .path(realm.name)
                .path("scim/v2/Users")
                .path(user.id)
                .build()
                .toString()
        )
    )

    data class ScimUserName(
        val familyName: String?,
        val givenName: String?
    )

    data class ScimUserEmail(
        val value: String?,
        val primary: Boolean
    )

    data class ScimUserMeta(
        val resourceType: String,
        val location: String
    )
}